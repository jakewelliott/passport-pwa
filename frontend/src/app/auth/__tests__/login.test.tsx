import { renderWithClient } from '@/lib/testing/test-wrapper';
import { fireEvent, screen, waitFor } from '@testing-library/react';
import { act } from 'react-dom/test-utils';
import { toast } from 'react-toastify';
import { beforeEach, describe, expect, it, vi } from 'vitest';
import LoginPage from '../login';

// Mock the react-toastify
vi.mock('react-toastify');

// Mock the auth hooks
const loginMutate = vi.fn();
const registerMutate = vi.fn();

vi.mock('@/hooks/auth/useLogin', () => ({
    useLogin: () => ({
        mutate: loginMutate,
    }),
}));

vi.mock('@/hooks/auth/useRegister', () => ({
    useRegister: () => ({
        mutate: registerMutate,
    }),
}));

const renderLoginPage = () => {
    renderWithClient(<LoginPage />);
};

describe('LoginPage', () => {
    beforeEach(() => {
        vi.clearAllMocks();
    });

    it('renders login form with all elements', () => {
        renderLoginPage();

        expect(screen.getByPlaceholderText('Username')).toBeInTheDocument();
        expect(screen.getByPlaceholderText('Password')).toBeInTheDocument();
        expect(screen.getByText('Login')).toBeInTheDocument();
        expect(screen.getByText('Register')).toBeInTheDocument();
    });

    it('shows validation errors when submitting empty form', async () => {
        renderLoginPage();

        fireEvent.click(screen.getByText('Login'));

        await waitFor(() => {
            expect(toast.error).toHaveBeenCalledWith('Username and Password are required.');
        });
    });

    it('shows validation errors when submitting empty field', async () => {
        renderLoginPage();
        fireEvent.change(screen.getByPlaceholderText('Username'), {
            target: { value: 'username' },
        });
        fireEvent.click(screen.getByText('Login'));

        await waitFor(() => {
            expect(toast.error).toHaveBeenCalledWith('Password is required.');
        });

        fireEvent.change(screen.getByPlaceholderText('Password'), {
            target: { value: 'password' },
        });
        fireEvent.change(screen.getByPlaceholderText('Username'), {
            target: { value: '' },
        });
        fireEvent.click(screen.getByText('Login'));

        await waitFor(() => {
            expect(toast.error).toHaveBeenCalledWith('Username is required.');
        });
    });

    it('handles successful login', async () => {
        renderLoginPage();

        fireEvent.change(screen.getByPlaceholderText('Username'), {
            target: { value: 'testuser' },
        });
        fireEvent.change(screen.getByPlaceholderText('Password'), {
            target: { value: 'password' },
        });

        await act(async () => {
            fireEvent.click(screen.getByText('Login'));
        });

        expect(loginMutate).toHaveBeenCalledWith(
            { username: 'testuser', password: 'password' },
            expect.objectContaining({
                onError: expect.any(Function),
            }),
        );
    });

    it('handles failed login', async () => {
        renderLoginPage();

        fireEvent.change(screen.getByPlaceholderText('Username'), {
            target: { value: 'wronguser' },
        });
        fireEvent.change(screen.getByPlaceholderText('Password'), {
            target: { value: 'wrongpass' },
        });

        let errorCallback: (error: Error) => void;
        loginMutate.mockImplementation((_, options) => {
            errorCallback = options.onError;
        });

        await act(async () => {
            fireEvent.click(screen.getByText('Login'));
            errorCallback?.(new Error('Invalid credentials'));
        });

        await waitFor(() => {
            expect(screen.getByPlaceholderText('Username')).toHaveClass('border-system_red');
            expect(screen.getByPlaceholderText('Username')).toHaveClass('ring-system_red');
        });
    });

    it('handles successful registration', async () => {
        renderLoginPage();

        fireEvent.change(screen.getByPlaceholderText('Username'), {
            target: { value: 'newuser' },
        });
        fireEvent.change(screen.getByPlaceholderText('Password'), {
            target: { value: 'password' },
        });

        await act(async () => {
            fireEvent.click(screen.getByText('Register'));
        });

        expect(registerMutate).toHaveBeenCalledWith(
            { username: 'newuser', password: 'password' },
            expect.objectContaining({
                onError: expect.any(Function),
            }),
        );
    });

    it('handles failed registration', async () => {
        renderLoginPage();

        fireEvent.change(screen.getByPlaceholderText('Username'), {
            target: { value: 'existinguser' },
        });
        fireEvent.change(screen.getByPlaceholderText('Password'), {
            target: { value: 'password' },
        });

        let errorCallback: (error: Error) => void;
        registerMutate.mockImplementation((_, options) => {
            errorCallback = options.onError;
        });

        await act(async () => {
            fireEvent.click(screen.getByText('Register'));
            errorCallback?.(new Error('Username already exists'));
        });

        await waitFor(() => {
            expect(screen.getByPlaceholderText('Username')).toHaveClass('border-system_red');
            expect(screen.getByPlaceholderText('Username')).toHaveClass('focus:border-system_red');
            expect(screen.getByPlaceholderText('Username')).toHaveClass('ring-system_red');
        });
    });

    it('renders privacy policy link', () => {
        renderLoginPage();

        const privacyLink = screen.getByText('Privacy Policy');
        expect(privacyLink).toBeInTheDocument();
        expect(privacyLink).toHaveAttribute('href', 'https://www.nc.gov/privacy');
        expect(privacyLink).toHaveAttribute('target', '_blank');
    });
});
