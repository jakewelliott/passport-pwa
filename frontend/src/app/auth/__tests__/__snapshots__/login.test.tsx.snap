// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LoginPage > matches snapshot 1`] = `
<div>
  <div
    class="flex h-full items-center justify-center bg-center bg-cover bg-no-repeat p-12"
    style="background-image: url("/photos/SILA-BackCover.jpg");"
  >
    <form
      class="flex flex-col items-center gap-3 rounded-3xl bg-supporting-lightblue bg-opacity-75 p-8"
    >
      <input
        class="bg-system-white w-full rounded-lg border p-3 focus:outline-none focus:ring-1 border-system-gray focus:border-secondary-darkteal focus:ring-secondary-darkteal"
        name="username"
        placeholder="Username"
        type="text"
      />
      <input
        class="bg-system-white w-full rounded-lg border p-3 focus:outline-none focus:ring-1 border-system-gray focus:border-secondary-darkteal focus:ring-secondary-darkteal"
        name="password"
        placeholder="Password"
        type="password"
      />
      <p
        class="text-center"
      >
        By using this application, you agree to the State of NC's
         
        <a
          class="text-secondary-darkteal hover:underline"
          href="https://www.nc.gov/privacy"
          rel="noopener noreferrer"
          target="_blank"
        >
          Privacy Policy
        </a>
      </p>
      <div
        class="flex flex-row gap-4"
      >
        <button
          class="flex cursor-pointer select-none items-center justify-center text-center rounded-full transition-all duration-200 p-2 w-48 bg-secondary-darkteal"
          type="button"
        >
          <p
            class="text-system-white text-2xl"
          >
            Register
          </p>
        </button>
        <button
          class="flex cursor-pointer select-none items-center justify-center text-center rounded-full transition-all duration-200 p-2 w-48 bg-secondary-orange"
          type="submit"
        >
          <p
            class="text-system-white text-2xl"
          >
            Login
          </p>
        </button>
      </div>
    </form>
  </div>
</div>
`;
